File: example/example-no-engine/src/main/java/io/holunda/camunda/bpm/data/example/noengine/CamundaBpmDataExampleNoEngineApplication.java
package io.holunda.camunda.bpm.data.example.noengine;

import io.holunda.camunda.bpm.data.CamundaBpmData;
import io.holunda.camunda.bpm.data.factory.VariableFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.ApplicationArguments;
import org.springframework.boot.ApplicationRunner;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;

@SpringBootApplication
public class CamundaBpmDataExampleNoEngineApplication implements ApplicationRunner {

  public static final VariableFactory<String> ORDER_ID = CamundaBpmData.stringVariable("orderId");
  public static final VariableFactory<Order> ORDER = CamundaBpmData.customVariable("order", Order.class);

  @Autowired
  private VariableManipulatingService component;

  @Override
  public void run(ApplicationArguments args) throws Exception {
    component.workWithVariables();
  }

  public static void main(String[] args) {
    SpringApplication.run(CamundaBpmDataExampleNoEngineApplication.class, args);
  }
}

File: example/example-no-engine/src/main/kotlin/CamundaBpmDataExampleNoEngineApplication.kt
package io.holunda.camunda.bpm.data.example.noengine

import io.holunda.camunda.bpm.data.CamundaBpmData
import io.holunda.camunda.bpm.data.CamundaBpmData.builder
import io.holunda.camunda.bpm.data.CamundaBpmData.reader
import io.holunda.camunda.bpm.data.CamundaBpmDataKotlin
import io.holunda.camunda.bpm.data.example.noengine.domain.Order
import io.holunda.camunda.bpm.data.example.noengine.domain.OrderPosition
import io.holunda.camunda.bpm.data.factory.VariableFactory
import mu.KLogging
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.ApplicationArguments
import org.springframework.boot.ApplicationRunner
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.runApplication
import org.springframework.stereotype.Component
import java.math.BigDecimal
import java.time.Instant.now
import java.util.*

fun main(args: Array<String>) = runApplication<CamundaBpmDataExampleNoEngineApplication>(*args).let { Unit }


val ORDER_ID = CamundaBpmData.stringVariable("orderId")
val ORDER: VariableFactory<Order> = CamundaBpmDataKotlin.customVariable("order")


@SpringBootApplication
class CamundaBpmDataExampleNoEngineApplication: ApplicationRunner {

  @Autowired
  lateinit var variableAwareComponent: VariableAwareComponent


  override fun run(args: ApplicationArguments) {
    variableAwareComponent.readAndWriteSomeVariables()
  }

}

@Component
class VariableAwareComponent {

  companion object: KLogging()

  fun readAndWriteSomeVariables(): Boolean {

    val position = OrderPosition("Pencil", BigDecimal.valueOf(10.00), 3)

    val vars = builder()
      .set(ORDER_ID, "4711")
      .set(ORDER, Order("4711", Date.from(now()), listOf(position)))
      .build()

    ORDER_ID.from(vars).get()

    val reader = reader(vars)
    val order = reader.get(ORDER)

    logger.info { "Successfully read $order" }

    return true
  }

}
